company.el

```diff
@@ -1558,54 +1537,66 @@ Only affects `company-text-icons-margin'."
   :type 'string)
 
 (defcustom company-text-icons-add-background nil
-  "When non-nil, add special background to the characters."
+  "When non-nil, generate a background color for text icons when none is given.
+See `company-text-icons-mapping'."
   :type 'boolean)
 
 (defun company-text-icons-margin (candidate selected)
   "Margin function which returns unicode icons."
   (when-let ((candidate candidate)
              (kind (company-call-backend 'kind candidate))
-             (icon (or (alist-get kind company-text-icons-mapping)
-                       (alist-get t company-text-icons-mapping)))
-             (face (or (assoc-default kind
-                                      company-text-kind-face-mapping)
-                       (assoc-default t company-text-kind-face-mapping))))
-    (propertize
-     (format company-text-icons-format icon)
-     'face
-     `(,@company-text-face-extra-attributes
-       ,@(when company-text-icons-add-background
-           (list :background
-                 (company-text-icons--background face selected)))
-       :foreground ,(face-attribute face :foreground)))))
+             (conf (or (alist-get kind company-text-icons-mapping)
+                       (alist-get t company-text-icons-mapping))))
+    (cl-destructuring-bind (icon &optional fg bg) conf
+      (propertize
+       (format company-text-icons-format icon)
+       'face
+       (company-text-icons--face fg bg selected)))))
 
 (declare-function color-rgb-to-hex "color")
 (declare-function color-gradient "color")
 
-(defun company-text-icons--background (face selected)
-  (apply #'color-rgb-to-hex
-         (nth 0 (color-gradient
-                 (color-name-to-rgb
-                  (face-attribute
-                   (if selected
-                       'company-tooltip-selection
-                     'company-tooltip)
-                   :background))
-                 (color-name-to-rgb
-                  (face-attribute face :foreground))
-                 10))))
+(defun company-text-icons--face (fg bg selected)
+  ;; Narrow to specific bg used for current candidate when a CONS cell.
+  (when (consp bg)
+    (setq bg (if selected (car bg) (cdr bg))))
```

Comment:
Perhaps swap `car` and `cdr` here? The non-selected case feels like the more important one, so it should go first.

Reply:
Done.